package PS;
import "proto/range.proto";

// information
message InstanceInfo {
  // TODO multi-label
  enum LabelType {
    // unsupvervised learning
    EMPTY = 1;
    // -1, 1
    BINARY = 2;
    // 0, 1, 2, ...
    MULTICLASS = 3;
    // regression
    CONTINUOUS = 4;
  }
  required LabelType label_type = 1;
  required FeatureGroupInfo all_group = 2;
  repeated FeatureGroupInfo individual_groups = 3;
}

message FeatureGroupInfo {
  required int32 group_id = 1;
  enum FeatureType {
    DENSE = 1;
    SPARSE = 2;
    SPARSE_BINARY = 3;
  }
  required FeatureType feature_type = 2;

  required uint64 feature_begin = 3 [default = 0xFFFFFFFFFFFFFFFF];
  required uint64 feature_end = 4;

  required uint64 num_instances = 5;
  required uint64 num_entries = 6;
}

message Instance {
  // -1, 1 for binary classification, 0,1,2,... for multiclass classfication,
  // real value for regression, and empty for unsupverised learning
  optional float label = 1;

  // three cases:
  // 1. dense: feature_id is emtpy
  // 2. sparse binary: value is emtpy
  // 3. sparse real-value: the size of feature should be equal to the size of value
  repeated int64 feature_id = 2;
  repeated float value = 3;
}
